### PR TITLE
Replace libsecp256k1 with secp256k1

### LINKED ISSUE #8089: Switch to rust-secp256k1 from libsecp256k1
It seems that currently we use `libsecp256k1` in `sp_core::ecdsa` implementation. As I've been told, the main reason for that was because `rust-secp256k1` wouldn't compile for WASM/no_std.

It seems it's [not the case any more](https://github.com/gregdhill/bench-secp256k1-wasm) and `rust-secp256k1` is more performant, has been audited and is more commonly used, hence I see no reasons for not switching to this.

### PR DESCRIPTION
**Pull request by davxy**
Closes #8089

[libsecp256k](https://crates.io/crates/libsecp256k1) has been replaced by [secp256k1](https://crates.io/crates/secp256k1).

According to the [benchmarks](https://github.com/davxy/crypto-benches) the latter has been found to be faster by a factor of three/four.

The interfaces are unchanged.

### CODE REVIEW

**Comment by ggwpez** [comment]
I re-ran your benchmarks on reference hardware and it looks great:  
```pre
k256/sign               time:   [71.721 us 71.765 us 71.822 us]                      
Found 8 outliers among 100 measurements (8.00%)
  1 (1.00%) low mild
  5 (5.00%) high mild
  2 (2.00%) high severe
k256/verify             time:   [86.915 us 87.016 us 87.163 us]                        
Found 14 outliers among 100 measurements (14.00%)
  1 (1.00%) low severe
  1 (1.00%) low mild
  4 (4.00%) high mild
  8 (8.00%) high severe

libsecp256k1/sign       time:   [96.030 us 96.106 us 96.195 us]                              
Found 9 outliers among 100 measurements (9.00%)
  3 (3.00%) high mild
  6 (6.00%) high severe
libsecp256k1/verify     time:   [129.83 us 129.88 us 129.93 us]                                
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe

secp256k1/sign          time:   [26.382 us 26.391 us 26.400 us]                            
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe
secp256k1/verify        time:   [27.856 us 27.865 us 27.875 us]                              
Found 10 outliers among 100 measurements (10.00%)
  6 (6.00%) high mild
  4 (4.00%) high severe
```
not much improvement when using our `production` profile:
```pre
k256/sign               time:   [67.466 us 67.503 us 67.547 us]                      
                        change: [-6.0147% -5.8332% -5.5328%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  5 (5.00%) high mild
  5 (5.00%) high severe
k256/verify             time:   [82.048 us 82.083 us 82.119 us]                        
                        change: [-5.8142% -5.6322% -5.4164%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) low severe
  5 (5.00%) high mild
  3 (3.00%) high severe

libsecp256k1/sign       time:   [96.219 us 96.269 us 96.326 us]                              
                        change: [+0.0031% +0.2746% +0.5675%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 6 outliers among 100 measurements (6.00%)
  3 (3.00%) high mild
  3 (3.00%) high severe
libsecp256k1/verify     time:   [129.47 us 129.51 us 129.56 us]                                
                        change: [-0.4219% -0.2543% -0.0827%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 7 outliers among 100 measurements (7.00%)
  2 (2.00%) high mild
  5 (5.00%) high severe

secp256k1/sign          time:   [26.367 us 26.378 us 26.392 us]                            
                        change: [-0.3535% -0.0463% +0.3154%] (p = 0.82 > 0.05)
                        No change in performance detected.
Found 7 outliers among 100 measurements (7.00%)
  3 (3.00%) high mild
  4 (4.00%) high severe
secp256k1/verify        time:   [27.293 us 27.303 us 27.314 us]                              
                        change: [-2.2911% -2.0530% -1.8383%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  4 (4.00%) high mild
  1 (1.00%) high severe
```

**Comment by davxy** [comment]
@ggwpez from what I see on your results for `production` profile:

libsecp256k1 (the old one):
- sign: 96.269 us
- verify: 129.51 us

secp256k1 (the new one):
- sign: 26.378 us 
- verify: 27.303 us

=> Signing has an improvement of 3.64
=> Verifying has an improvement of 4.74

Not bad ;-)

**Comment by bkchr** [comment]
@crystalin that will also make you happy ;)

**Comment by bkchr** [comment]
@davxy this requires a companion in Polkadot.

**Comment by burdges** [comment]
We do not have that many key pairs, so maybe we could just `Arc` them all and never move them?

**Comment by bkchr** [comment]
> We do not have that many key pairs, so maybe we could just `Arc` them all and never move them?

Can you be more specific? What you mean?

**Comment by burdges** [comment]
I'm wondering if keypairs could exist in one place, and never really be moved, except internally by cryptographic code. 

I also realized the answer gets messy since I'd wanted to use key splitting in the BEEFY BLS key, which makes the secret key mutable.

**Comment by bkchr** [comment]
> I'm wondering if keypairs could exist in one place, and never really be moved, except internally by cryptographic code.

Keypairs are normally in the keystore and don't move around.

**Comment by athei** [comment]
This doesn't replace `libsecp256k1` completely. It is still in most manifests. Is this intended?

**Comment by davxy** [comment]
> This doesn't replace `libsecp256k1` completely. It is still in most manifests. Is this intended?

@athei yes is still there. In particular it is still used for:
* recovery of public key when the signature is not "standard" (i.e. signature larger than curve order, this is handled by `libsecp256k1` only).
* Within FRAME in a couple of very circumscribed (can be optionally easily replaced on both): 
    - By `beefy-mmr` to convert ecdsa PublicKey to ETH address [here](https://github.com/paritytech/substrate/blob/c8bef76781fad6e88aa85397ece5058db780f12c/frame/beefy-mmr/src/lib.rs#L73-L91)
    - By a `contracts` benchmark [here](https://github.com/paritytech/substrate/blob/c8bef76781fad6e88aa85397ece5058db780f12c/frame/contracts/src/benchmarking/mod.rs#L1878-L1885)

**Comment by bkchr** [comment]
> * recovery of public key when the signature is not "standard" (i.e. signature larger than curve order, this is handled by `libsecp256k1` only).
> 

This requires that we check on all possible chains if they can still sync, if we disable the old broken code.

> * Within FRAME in a couple of very circumscribed places (can be optionally easily replaced on both):
>   
>   * By `beefy-mmr` to convert ecdsa PublicKey to ETH address [here](https://github.com/paritytech/substrate/blob/c8bef76781fad6e88aa85397ece5058db780f12c/frame/beefy-mmr/src/lib.rs#L73-L91)
>   * By a `contracts` benchmark [here](https://github.com/paritytech/substrate/blob/c8bef76781fad6e88aa85397ece5058db780f12c/frame/contracts/src/benchmarking/mod.rs#L1878-L1885)

These two should be replaced.

**Comment by davxy** [comment]
# CI BUILD ISSUE

## TL;DR: 

Issues building for wasm32 on the CI container built from Debian 10. It contains clang v7.
Latest Debian stable (11) ships with clang v11. Where the problem is not present.

## Problem Analysis

As can be see [here](https://gitlab.parity.io/parity/substrate/-/jobs/1374013) the polkadot companion fails with this error:

```
    running: "sccache" "clang" "-O3" "-ffunction-sections" "-fdata-sections" "-fPIC" "--target=wasm32-unknown-unknown" "-I" "depend/secp256k1/" "-I" "depend/secp256k1/include" "-I" "depend/secp256k1/src" "-I" "wasm-sysroot" "-Wall" "-Wextra" "-DSECP256K1_API=" "-DENABLE_MODULE_ECDH=1" "-DENABLE_MODULE_SCHNORRSIG=1" "-DENABLE_MODULE_EXTRAKEYS=1" "-DUSE_NUM_NONE=1" "-DUSE_FIELD_INV_BUILTIN=1" "-DUSE_SCALAR_INV_BUILTIN=1" "-DECMULT_GEN_PREC_BITS=4" "-DECMULT_WINDOW_SIZE=15" "-DUSE_EXTERNAL_DEFAULT_CALLBACKS=1" "-o" "/builds/parity/substrate/companions/polkadot/target/debug/wbuild/kusama-runtime/target/wasm32-unknown-unknown/release/build/secp256k1-sys-a327a13c1bed59b7/out/depend/secp256k1/src/secp256k1.o" "-c" "depend/secp256k1/src/secp256k1.c"
    cargo:warning=In file included from depend/secp256k1/src/secp256k1.c:12:
    cargo:warning=In file included from depend/secp256k1/src/assumptions.h:10:
    cargo:warning=In file included from /usr/lib/llvm-7/lib/clang/7.0.1/include/limits.h:37:
    cargo:warning=/usr/include/limits.h:26:10: fatal error: 'bits/libc-header-start.h' file not found
    cargo:warning=#include <bits/libc-header-start.h>
    cargo:warning=         ^~~~~~~~~~~~~~~~~~~~~~~~~~
    cargo:warning=1 error generated.
    exit status: 1
    --- stderr
```

After some investigation I've reduced the error to this simple C program:

```
#include <limits.h>           // also included by secp256k1.c
int main() { return 0; }
```

Trying to build this program with the `paritytech/ci-linux:production` container (the one used for CI) I get the same error:

```
root@7119bae7d14f:/builds# clang -c -v --target=wasm32-unknown-unknown poc.c
clang version 7.0.1-8+deb10u2 (tags/RELEASE_701/final)
Target: wasm32-unknown-unknown
...
#include "..." search starts here:
#include <...> search starts here:          <===  NOTE THE AVAILABLE INCLUDE PATHS SEARCH and their ORDER
 /usr/local/include
 /usr/lib/llvm-7/lib/clang/7.0.1/include    <======== THIS contains <limits.h>
 /usr/include                               <======== THIS also contains <limits.h>
End of search list.
In file included from poc.c:1:
In file included from /usr/lib/llvm-7/lib/clang/7.0.1/include/limits.h:37:
/usr/include/limits.h:26:10: fatal error: 'bits/libc-header-start.h' file not found
#include <bits/libc-header-start.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~~~
```

The problem arises from an `include` directive in `/usr/lib/llvm-7/lib/clang/7.0.1/include/limits.h:36`:

```
#if __STDC_HOSTED__ && __has_include_next(<limits.h>)
#include_next <limits.h>
#end
```
Now, given that:
- `__STDC_HOSTED__ = 1` and
- `__has_include_next(<limits.h>)`  is true because `/usr/include` contains`<limits.h>` as well

Follows that the `/usr/inlcude/limits.h` is included.

But this header file contains the inclusion of `<bits/libc-header-start.h>` and that is not in any of the previous search paths.

When instead we build for `x86-64` target the additional `/usr/include/x86_64-linux-gnu` path is present in the list (and that is where the `libc-header-start` file is).

### CLANG-7 Issue

Long story short, the issue is because of clang version.

The CI image has been built from debian 10 (buster). It ships with clang v7.

The issue is not present in the latest debian stable 11. It ships with clang v11.


This is the output when building with clang 11:

```
root@8bbd9b945e1a:/# clang -c -v --target=wasm32-unknown-unknown poc.c
Debian clang version 11.0.1-2
Target: wasm32-unknown-unknown
...
#include "..." search starts here:
#include <...> search starts here:
 /usr/lib/llvm-11/lib/clang/11.0.1/include         <======= Note single <limits.h>
End of search list.
```

Here the `__has_include_next(<limits.h>)` check fails and thus the build succeeded as expected.

**Comment by athei** [comment]
You changed beefy to use secp256k1 within the runtime. There is no precedence for building C code to wasm within substrate. I think we want to keep it this way. Cross compiling with C isn't so easy as with Rust. Most operating system won't have a toolchain installed out of the box.

So we should either keep `secp256k1` for beefy or make it use an existing host function to decompress the key (if there is one).

**Comment by koute** [comment]
> Cross compiling C isn't so easy as Rust. Most operating systems won't have a toolchain installed out of the box.
> 
> So we should either keep `secp256k1` for beefy or make it use an existing host function to decompress the key (if there is one).

Yes, I agree. Let's not open that can of worms; if this is needed in WASM we should definitely either use a Rust-only implementation, or preferably just do it through a host function. (Which is most likely going to be the fastest anyway.)

**Comment by athei** [comment]
Benchmarks are also compiled to wasm. So I guess the contracts change needs the same treatment as beefy.

**Comment by davxy** [comment]
In the end, looks like for the moment is better to stick with the `libsecp256k1` for stuff expected to be built in wasm. 
(furthermore this allows to don't depend on CI build machine upgrade [this](https://github.com/paritytech/ci_cd/issues/324))

Once that the `k256` pure Rust implementation will improve performances we can port everything to `k256` (both native and wasm). @tarcieri :-)

Nevertheless the port for `sp-io` and `sp-core` has been achieved correctly.

**Comment by tarcieri** [comment]
@davxy curious why you wouldn't move to `k256` for WASM. The performance is still better than libsecp256k1

**Comment by bkchr** [comment]
@tarcieri Verifying is still more than factor 2 faster than with `k256`. I mean we would switch instantly to `k256`, but we already have "problems" with ecdsa being slow. So, we would like to use the fastest option. But, if you tell me how we can help you to make it faster, I would be open to also switch to `k256`. :)

**Comment by tarcieri** [comment]
@bkchr huh? per the benchmarks above and my own benchmarks, `k256` is significantly faster at both signing and verification. Are you talking about WASM-specific benchmarks or something, and if so, where are those? (I would still be surprised if that changed anyting)

Here are [@ggwpez benchmarks from above, formatted into a table](https://github.com/paritytech/substrate/pull/10798#issuecomment-1030131130):

| crate              | profile       |  op     | median (μs) |
|--------------|-----------|-------|-------------|
| secp256k1     | release(?) | sign   | 26.4              |
| secp256k1     | production | sign  | 26.4             |
| secp256k1     | release(?) | verify | 27.9              |
| secp256k1     | production | verify | 27.3            |
| k256               | release(?) | sign   | 71.8              |
| k256               | production | sign  | 67.5             |
| k256               | release(?) | verify | 87.0             |
| k256               | production | verify | 82.0           |
| libsecp256k1 | release(?) | sign    | 96.1             |
| libsecp256k1 | production | sign   | 96.3            |
| libsecp256k1 | release(?) | verify  | 130.0          |
| libsecp256k1 | production | verify | 130.0         |

**Comment by bkchr** [comment]
@tarcieri Sorry, I should have clarified that I meant `secp256k1` vs `k256`.

**Comment by tarcieri** [comment]
@bkchr haha, no worries, the names are so confusing, especially since the C library is named `libsecp256k1`

**Comment by bkchr** [comment]
Is there any way we could support you @tarcieri to improve the performance of `k256`? Verify would be the most important part for us :)

**Comment by davxy** [comment]
My opinion is to:

- pick up the `secp256k1` fruits right now for the native code (as we did in this PR).
- start replacing with `k256` everything that is still using `libsecp256` right now. There is already a performance benefit using it as compared to `libsecp256`.
- and, as all of you already said, once the gap is less important we can switch everything to `k256`.

So, if for you is worth, I would like to finalize the PR by porting the stuff that is still using `libsecp256` to `k256`.

I also continue to monitor the improvements on the `k256` library. I'll set up something doing this automatically from time to time using the benches

**Comment by bkchr** [comment]
Sounds good :) 

Regarding the benchmarking code, I would before check again if `sp-io` not already supports the use cases.

**Comment by athei** [comment]
> Sorry @athei, what is your opinion? :D

Hehe no worries. I think I just followed up with my opinion. I think @davxy plan make sense.

**Comment by bkchr** [comment]
> It is only used to generate random keys there. I just checked that there is no host function for key generation.

There is one ;) https://github.com/paritytech/substrate/blob/master/primitives/application-crypto/src/traits.rs#L108

File: frame/beefy-mmr/Cargo.toml [hunk 1]
```diff
@@ -9,8 +9,8 @@ repository = "https://github.com/paritytech/substrate"
 
 [dependencies]
 hex = { version = "0.4", optional = true }
-codec = { version = "2.2.0", package = "parity-scale-codec", default-features = false, features = ["derive"] }
 libsecp256k1 = { version = "0.7.0", default-features = false }
+codec = { version = "2.2.0", package = "parity-scale-codec", default-features = false, features = ["derive"] }
```
Comment (athei) [inline frame/beefy-mmr/Cargo.toml:13]: Unrelated change.

File: frame/beefy-mmr/src/lib.rs [hunk 2]
```diff
@@ -70,13 +70,13 @@ where
...
 pub struct BeefyEcdsaToEthereum;
+
 impl Convert<beefy_primitives::crypto::AuthorityId, Vec<u8>> for BeefyEcdsaToEthereum {
 	fn convert(a: beefy_primitives::crypto::AuthorityId) -> Vec<u8> {
 		use sp_core::crypto::ByteArray;
-		let compressed_key = a.as_slice();
 
 		libsecp256k1::PublicKey::parse_slice(
-			compressed_key,
+			a.as_slice(),
```
Comment (athei) [inline frame/beefy-mmr/src/lib.rs:79]: Unrelated change.

File: frame/contracts/src/benchmarking/mod.rs [hunk 3]
```diff
@@ -1922,13 +1922,13 @@ benchmarks! {
...
-		let mut rng = rand_pcg::Pcg32::seed_from_u64(123456);
+		use libsecp256k1::{SecretKey, Message, sign};
 
+		let mut rng = rand_pcg::Pcg32::seed_from_u64(123456);
 		let message_hash = sp_io::hashing::blake2_256("Hello world".as_bytes());
+
 		let signatures = (0..r * API_BENCHMARK_BATCH_SIZE)
 			.map(|i| {
-				use libsecp256k1::{SecretKey, Message, sign};
-
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:1931]: Unrelated change.

**Review by athei — COMMENTED** [review]
I think we should merge this without any changes to MMR and contracts. We can move them over to k256 or host functions on another PR. Here we should just reap the performance benefits in our host paths.

**Comment by davxy** [comment]
> I think we should merge this without any changes to MMR and contracts. We can move them over to k256 or host functions on another PR. Here we should just reap the performance benefits in our host paths.

For me is good as well.

@bkchr @athei if you accept this PR then I'll start working on another PR for `libsecp256` replacement.

**Review by athei — APPROVED** [review]

**Comment by rcny** [comment]
`paritytech/ci-linux:production` image is bundled with `clang-13` now.

File: primitives/io/src/lib.rs [hunk 4]
```diff
@@ -920,16 +928,16 @@ pub trait Crypto {
...
-		)
-		.map_err(|_| EcdsaVerifyError::BadV)?;
-		let pubkey = libsecp256k1::recover(&libsecp256k1::Message::parse(msg), &rs, &v)
+		let msg = Message::from_slice(msg).expect("Message is 32 bytes; qed");
+		let pubkey = SECP256K1
+			.recover_ecdsa(&msg, &sig)
 			.map_err(|_| EcdsaVerifyError::BadSignature)?;
 		let mut res = [0u8; 64];
-		res.copy_from_slice(&pubkey.serialize()[1..65]);
+		res.copy_from_slice(&pubkey.serialize_uncompressed()[1..]);
```
[the remark this answers is not in the document]
Comment (davxy) [reply primitives/io/src/lib.rs:942]: You got it right. The old version handles, non standard, overflowing signatures.
I've added some notes to the docs

File: primitives/io/src/lib.rs [hunk 5]
```diff
@@ -895,14 +902,15 @@ pub trait Crypto {
...
+			if sig[64] > 26 { sig[64] - 27 } else { sig[64] } as u8,
 		)
 		.map_err(|_| EcdsaVerifyError::BadV)?;
-		let pubkey = libsecp256k1::recover(&libsecp256k1::Message::parse(msg), &rs, &v)
-			.map_err(|_| EcdsaVerifyError::BadSignature)?;
+		let sig = libsecp256k1::Signature::parse_overflowing_slice(&sig[..64])
+			.map_err(|_| EcdsaVerifyError::BadRS)?;
+		let msg = libsecp256k1::Message::parse(msg);
+		let pubkey =
+			libsecp256k1::recover(&msg, &sig, &rid).map_err(|_| EcdsaVerifyError::BadSignature)?;
```
[the remark this answers is not in the document]
Comment (davxy) [reply primitives/io/src/lib.rs:915]: Actually I've preferred to rename the variable from `v` to `rid` for coherence to the common jargon.

`v` is the name commonly used to refer to the "recovery id before the subtraction of the  constant value (27)".

See also [here](https://bitcoin.stackexchange.com/questions/83035/how-to-determine-first-byte-recovery-id-for-signatures-message-signing)

File: primitives/io/src/lib.rs [hunk 6]
```diff
@@ -943,14 +951,15 @@ pub trait Crypto {
...
+			if sig[64] > 26 { sig[64] - 27 } else { sig[64] } as u8,
 		)
 		.map_err(|_| EcdsaVerifyError::BadV)?;
-		let pubkey = libsecp256k1::recover(&libsecp256k1::Message::parse(msg), &rs, &v)
-			.map_err(|_| EcdsaVerifyError::BadSignature)?;
+		let sig = libsecp256k1::Signature::parse_overflowing_slice(&sig[0..64])
+			.map_err(|_| EcdsaVerifyError::BadRS)?;
+		let msg = libsecp256k1::Message::parse(msg);
+		let pubkey =
+			libsecp256k1::recover(&msg, &sig, &rid).map_err(|_| EcdsaVerifyError::BadSignature)?;
```
[the remark this answers is not in the document]
Comment (davxy) [reply primitives/io/src/lib.rs:964]: Same as above

**Review by wigy-opensource-developer — APPROVED** [review]
Nice work. I have just added one nitpicking comment, but would be fine without that, too.

File: primitives/core/src/ecdsa.rs [hunk 7]
```diff
@@ -507,57 +502,57 @@ impl Pair {
...
-	let ri = libsecp256k1::RecoveryId::parse(x[64])?;
-	Ok((sig, ri))
+impl Drop for Pair {
+	fn drop(&mut self) {
+		let ptr = self.secret.as_mut_ptr();
+		for off in 0..self.secret.len() {
+			unsafe {
+				core::ptr::write_volatile(ptr.add(off), 0);
+			}
+		}
```
[the remark this answers is not in the document]
Comment (tarcieri) [reply primitives/core/src/ecdsa.rs:554]: It looks like `zeroize` is already a dependency. You could use it here:

```suggestion
		use zeroize::Zeroize;
		self.secret.zeroize();
```

[the remark this answers is not in the document]
Comment (davxy) [reply primitives/core/src/ecdsa.rs:554]: Unfortunately the `secp256k1::SecretKey` doesn't implement `DefaultIsZeroes` nor `Zeroize` traits :-/

[the remark this answers is not in the document]
Comment (davxy) [reply primitives/core/src/ecdsa.rs:554]: > Nice touch, but https://doc.rust-lang.org/core/ptr/fn.write_bytes.html might perform better on some platforms and is also more expressive. Also, have you considered `Seed` is also an important secret?

`write_volatile` is less performant but it is guaranteed to not be elided or reordered by the compiler especially with optimizations enabled

**Review by bkchr — APPROVED** [review]
We should do a "burn in" before merging this finally.

Syncing once Kusama and Polkadot from the start to tip would be good.

**Comment by davxy** [comment]
bot merge

**Comment by davxy** [comment]
bot merge

**Comment by athei** [comment]
Doesn't we need to re-benchmark the runtime? I assume at least some values are depended on this. Like block base weight or something?

We cannot get any more throughput without these improvements being reflected in runtime benchmarks, right? I am wondering how we do capitalize on these improvements? Do we tune some numbers of the parachain protocol? Can someone explain?

**Comment by bkchr** [comment]
We regenerate weights when doing a new release for Polkadot/Kusama.

Yes that should increase the throughput of Moonbeam for example. Their tx base weight should probably decrease because of this change, assuming we have signature verification being part of this benchmark.

File: primitives/core/src/ecdsa.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Slesarew) [reply primitives/core/src/ecdsa.rs:554]: Is there a test for this cleanup anywhere?
